### PR TITLE
Further improvements to experiments matrix

### LINKED ISSUE #64: Implement extra custom experiment matrix views
Replicate the WL1 buttons at the top of the matrix view for logged-in users, i.e.
* Show just moderated / all public / all visible-to-me / my experiments
* Extra buttons to hide moderated models / protocols from views that include these among others (all the above views include moderated entities by default)

The default view for anonymous users should be moderated experiments (which requires #41 & requires #40). The default for for logged-in users is 'my experiments' (which includes moderated ones as well by default).

### LINKED ISSUE #73: Review speed of experiments matrix view
As we get more experiments on scrambler, we should check it's presenting the matrix at least as quickly as WL1 could.

Child of #68.

### PR DESCRIPTION
**Pull request by jonc125**
Fixes #64. Fixes #73.

### CODE REVIEW

**Comment by jonc125** [comment]
As mentioned on the issue, we can now generate the default matrix view on scrambler in <1s. There's probably still some scope for improvement, but I don't think it's worth me spending more time on it now! We can review again if it gets annoyingly slow in the future.

**Comment by jonc125** [comment]
Given this is a reasonably significant change I'd appreciate a thorough code review before merging.

File: weblab/core/visibility.py [hunk 1]
```diff
@@ -19,23 +19,38 @@ class Visibility:
...
     """
-    # Ordered by most conservative first
-    levels = [
-        Visibility.PRIVATE, Visibility.PUBLIC, Visibility.MODERATED,
-    ]
-
     # Use the most conservative of the two entities' visibilities
-    return levels[min(
-        levels.index(vis)
+    return LEVELS[min(
```
Comment (MichaelClerx) [inline weblab/core/visibility.py]: Surprised you don't need `min([...])` instead of `min(...)`, didn't know that!

**Comment by jonc125** [comment]
> Having some trouble deploying this

Given it's failing an assertion within `self._repo.head.is_valid()` it sounds like one of your model or protocol git repos is really mucked up! At a guess from the debug lines, entity id 19. Can you try deleting that model/protocol then re-deploy?

**Comment by MichaelClerx** [comment]
I went a bit more drastic than that and did a `vagrant destory`. Back up and running now, but having trouble adding new users to test the private / public cases. I get a gateway timeout trying to register, and adding a user via the django back-end also timed out. Could this be anything to do with the DB? Or more likely an email issue?

**Comment by MichaelClerx** [comment]
User _was_ created, despite gateway timeout it seems

**Comment by jonc125** [comment]
It'll be email sending I think - it needs to send emails so the person registering can verify their account, and then so the admin is alerting and can give them permissions.

**Comment by MichaelClerx** [comment]
Just tried switching my (very new) system back to `master` again (by changing the django branch in the Vagrantfile), and got this:

```
fatal: [localhost]: FAILED! => {
  "changed": false,
  "cmd": "./manage.py migrate --noinput --settings=config.settings.deployed",
  "msg":         ____________MESSAGE_________________,
  "path": "/opt/django/venv/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin",
  "state": "absent",
  "syspath": [
    "/tmp/ansible_lqElCa",
    "/tmp/ansible_lqElCa/ansible_modlib.zip",
    "/tmp/ansible_lqElCa/ansible_modlib.zip",
    "/usr/lib/python2.7",
    "/usr/lib/python2.7/plat-x86_64-linux-gnu",
    "/usr/lib/python2.7/lib-tk",
    "/usr/lib/python2.7/lib-old",
    "/usr/lib/python2.7/lib-dynload",
    "/usr/local/lib/python2.7/dist-packages",
    "/usr/lib/python2.7/dist-packages"
  ]
}
```

```
________________MESSAGE________________________
stdout: DEBUG 2019-05-28 16:06:02,530 git.cmd: Popen(['git', 'version'], cwd=/opt/django/WebLab/weblab, universal_newlines=False, shell=None)
DEBUG 2019-05-28 16:06:02,534 git.cmd: Popen(['git', 'version'], cwd=/opt/django/WebLab/weblab, universal_newlines=False, shell=None)
DEBUG 2019-05-28 16:06:02,675 django.db.backends: (0.002) 
            SELECT c.relname, c.relkind
            FROM pg_catalog.pg_class c
            LEFT JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
            WHERE c.relkind IN ('r', 'v')
                AND n.nspname NOT IN ('pg_catalog', 'pg_toast')
                AND pg_catalog.pg_table_is_visible(c.oid); args=None
DEBUG 2019-05-28 16:06:02,678 django.db.backends: (0.001) SELECT \"django_migrations\".\"app\", \"django_migrations\".\"name\" FROM \"django_migrations\"; args=()
DEBUG 2019-05-28 16:06:02,681 django.db.backends: (0.001) 
            SELECT c.relname, c.relkind
            FROM pg_catalog.pg_class c
            LEFT JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
            WHERE c.relkind IN ('r', 'v')
                AND n.nspname NOT IN ('pg_catalog', 'pg_toast')
                AND pg_catalog.pg_table_is_visible(c.oid); args=None
DEBUG 2019-05-28 16:06:02,683 django.db.backends: (0.000) SELECT \"django_migrations\".\"app\", \"django_migrations\".\"name\" FROM \"django_migrations\"; args=()
Operations to perform:
  Apply all migrations: accounts, admin, auth, contenttypes, entities, experiments, guardian, repocache, sessions, social_django
Running migrations:
  No migrations to apply.
DEBUG 2019-05-28 16:06:02,807 django.db.backends: (0.001) 
            SELECT c.relname, c.relkind
            FROM pg_catalog.pg_class c
            LEFT JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
            WHERE c.relkind IN ('r', 'v')
                AND n.nspname NOT IN ('pg_catalog', 'pg_toast')
                AND pg_catalog.pg_table_is_visible(c.oid); args=None
DEBUG 2019-05-28 16:06:02,808 django.db.backends: (0.000) SELECT \"django_migrations\".\"app\", \"django_migrations\".\"name\" FROM \"django_migrations\"; args=()
DEBUG 2019-05-28 16:06:02,811 django.db.backends: (0.001) SELECT \"django_content_type\".\"id\", \"django_content_type\".\"app_label\", \"django_content_type\".\"model\" FROM \"django_content_type\" WHERE (\"django_content_type\".\"model\" = 'logentry' AND \"django_content_type\".\"app_label\" = 'admin'); args=('logentry', 'admin')
DEBUG 2019-05-28 16:06:02,815 django.db.backends: (0.001) SELECT \"auth_permission\".\"content_type_id\", \"auth_permission\".\"codename\" FROM \"auth_permission\" INNER JOIN \"django_content_type\" ON (\"auth_permission\".\"content_type_id\" = \"django_content_type\".\"id\") WHERE \"auth_permission\".\"content_type_id\" IN (1) ORDER BY \"django_content_type\".\"app_label\" ASC, \"django_content_type\".\"model\" ASC, \"auth_permission\".\"codename\" ASC; args=(1,)
DEBUG 2019-05-28 16:06:02,816 django.db.backends: (0.000) SELECT \"django_content_type\".\"id\", \"django_content_type\".\"app_label\", \"django_content_type\".\"model\" FROM \"django_content_type\" WHERE \"django_content_type\".\"app_label\" = 'admin'; args=('admin',)
DEBUG 2019-05-28 16:06:02,818 django.db.backends: (0.000) SELECT \"django_content_type\".\"id\", \"django_content_type\".\"app_label\", \"django_content_type\".\"model\" FROM \"django_content_type\" WHERE (\"django_content_type\".\"model\" = 'group' AND \"django_content_type\".\"app_label\" = 'auth'); args=('group', 'auth')
DEBUG 2019-05-28 16:06:02,819 django.db.backends: (0.001) SELECT \"django_content_type\".\"id\", \"django_content_type\".\"app_label\", \"django_content_type\".\"model\" FROM \"django_content_type\" WHERE (\"django_content_type\".\"model\" = 'permission' AND \"django_content_type\".\"app_label\" = 'auth'); args=('permission', 'auth')
DEBUG 2019-05-28 16:06:02,822 django.db.backends: (0.001) SELECT \"auth_permission\".\"content_type_id\", \"auth_permission\".\"codename\" FROM \"auth_permission\" INNER JOIN \"django_content_type\" ON (\"auth_permission\".\"content_type_id\" = \"django_content_type\".\"id\") WHERE \"auth_permission\".\"content_type_id\" IN (2, 3) ORDER BY \"django_content_type\".\"app_label\" ASC, \"django_content_type\".\"model\" ASC, \"auth_permission\".\"codename\" ASC; args=(2, 3)
DEBUG 2019-05-28 16:06:02,823 django.db.backends: (0.001) SELECT \"django_content_type\".\"id\", \"django_content_type\".\"app_label\", \"django_content_type\".\"model\" FROM \"django_content_type\" WHERE \"django_content_type\".\"app_label\" = 'auth'; args=('auth',)
DEBUG 2019-05-28 16:06:02,825 django.db.backends: (0.000) SELECT \"django_content_type\".\"id\", \"django_content_type\".\"app_label\", \"django_content_type\".\"model\" FROM \"django_content_type\" WHERE (\"django_content_type\".\"model\" = 'contenttype' AND \"django_content_type\".\"app_label\" = 'contenttypes'); args=('contenttype', 'contenttypes')
DEBUG 2019-05-28 16:06:02,827 django.db.backends: (0.001) SELECT \"auth_permission\".\"content_type_id\", \"auth_permission\".\"codename\" FROM \"auth_permission\" INNER JOIN \"django_content_type\" ON (\"auth_permission\".\"content_type_id\" = \"django_content_type\".\"id\") WHERE \"auth_permission\".\"content_type_id\" IN (4) ORDER BY \"django_content_type\".\"app_label\" ASC, \"django_content_type\".\"model\" ASC, \"auth_permission\".\"codename\" ASC; args=(4,)
DEBUG 2019-05-28 16:06:02,829 django.db.backends: (0.000) SELECT \"django_content_type\".\"id\", \"django_content_type\".\"app_label\", \"django_content_type\".\"model\" FROM \"django_content_type\" WHERE \"django_content_type\".\"app_label\" = 'contenttypes'; args=('contenttypes',)
DEBUG 2019-05-28 16:06:02,830 django.db.backends: (0.000) SELECT \"django_content_type\".\"id\", \"django_content_type\".\"app_label\", \"django_content_type\".\"model\" FROM \"django_content_type\" WHERE (\"django_content_type\".\"model\" = 'session' AND \"django_content_type\".\"app_label\" = 'sessions'); args=('session', 'sessions')
DEBUG 2019-05-28 16:06:02,832 django.db.backends: (0.001) SELECT \"auth_permission\".\"content_type_id\", \"auth_permission\".\"codename\" FROM \"auth_permission\" INNER JOIN \"django_content_type\" ON (\"auth_permission\".\"content_type_id\" = \"django_content_type\".\"id\") WHERE \"auth_permission\".\"content_type_id\" IN (5) ORDER BY \"django_content_type\".\"app_label\" ASC, \"django_content_type\".\"model\" ASC, \"auth_permission\".\"codename\" ASC; args=(5,)
DEBUG 2019-05-28 16:06:02,833 django.db.backends: (0.000) SELECT \"django_content_type\".\"id\", \"django_content_type\".\"app_label\", \"django_content_type\".\"model\" FROM \"django_content_type\" WHERE \"django_content_type\".\"app_label\" = 'sessions'; args=('sessions',)
DEBUG 2019-05-28 16:06:02,835 django.db.backends: (0.001) SELECT \"django_content_type\".\"id\", \"django_content_type\".\"app_label\", \"django_content_type\".\"model\" FROM \"django_content_type\" WHERE (\"django_content_type\".\"model\" = 'code' AND \"django_content_type\".\"app_label\" = 'social_django'); args=('code', 'social_django')
DEBUG 2019-05-28 16:06:02,836 django.db.backends: (0.000) SELECT \"django_content_type\".\"id\", \"django_content_type\".\"app_label\", \"django_content_type\".\"model\" FROM \"django_content_type\" WHERE (\"django_content_type\".\"model\" = 'nonce' AND \"django_content_type\".\"app_label\" = 'social_django'); args=('nonce', 'social_django')
DEBUG 2019-05-28 16:06:02,838 django.db.backends: (0.000) SELECT \"django_content_type\".\"id\", \"django_content_type\".\"app_label\", \"django_content_type\".\"model\" FROM \"django_content_type\" WHERE (\"django_content_type\".\"model\" = 'association' AND \"django_content_type\".\"app_label\" = 'social_django'); args=('association', 'social_django')
DEBUG 2019-05-28 16:06:02,839 django.db.backends: (0.000) SELECT \"django_content_type\".\"id\", \"django_content_type\".\"app_label\", \"django_content_type\".\"model\" FROM \"django_content_type\" WHERE (\"django_content_type\".\"model\" = 'partial' AND \"django_content_type\".\"app_label\" = 'social_django'); args=('partial', 'social_django')
DEBUG 2019-05-28 16:06:02,841 django.db.backends: (0.000) SELECT \"django_content_type\".\"id\", \"django_content_type\".\"app_label\", \"django_content_type\".\"model\" FROM \"django_content_type\" WHERE (\"django_content_type\".\"model\" = 'usersocialauth' AND \"django_content_type\".\"app_label\" = 'social_django'); args=('usersocialauth', 'social_django')
DEBUG 2019-05-28 16:06:02,843 django.db.backends: (0.001) SELECT \"auth_permission\".\"content_type_id\", \"auth_permission\".\"codename\" FROM \"auth_permission\" INNER JOIN \"django_content_type\" ON (\"auth_permission\".\"content_type_id\" = \"django_content_type\".\"id\") WHERE \"auth_permission\".\"content_type_id\" IN (8, 9, 10, 6, 7) ORDER BY \"django_content_type\".\"app_label\" ASC, \"django_content_type\".\"model\" ASC, \"auth_permission\".\"codename\" ASC; args=(8, 9, 10, 6, 7)
DEBUG 2019-05-28 16:06:02,844 django.db.backends: (0.000) SELECT \"django_content_type\".\"id\", \"django_content_type\".\"app_label\", \"django_content_type\".\"model\" FROM \"django_content_type\" WHERE \"django_content_type\".\"app_label\" = 'social_django'; args=('social_django',)
DEBUG 2019-05-28 16:06:02,846 django.db.backends: (0.000) SELECT \"django_content_type\".\"id\", \"django_content_type\".\"app_label\", \"django_content_type\".\"model\" FROM \"django_content_type\" WHERE (\"django_content_type\".\"model\" = 'groupobjectpermission' AND \"django_content_type\".\"app_label\" = 'guardian'); args=('groupobjectpermission', 'guardian')
DEBUG 2019-05-28 16:06:02,847 django.db.backends: (0.000) SELECT \"django_content_type\".\"id\", \"django_content_type\".\"app_label\", \"django_content_type\".\"model\" FROM \"django_content_type\" WHERE (\"django_content_type\".\"model\" = 'userobjectpermission' AND \"django_content_type\".\"app_label\" = 'guardian'); args=('userobjectpermission', 'guardian')
DEBUG 2019-05-28 16:06:02,849 django.db.backends: (0.001) SELECT \"auth_permission\".\"content_type_id\", \"auth_permission\".\"codename\" FROM \"auth_permission\" INNER JOIN \"django_content_type\" ON (\"auth_permission\".\"content_type_id\" = \"django_content_type\".\"id\") WHERE \"auth_permission\".\"content_type_id\" IN (11, 12) ORDER BY \"django_content_type\".\"app_label\" ASC, \"django_content_type\".\"model\" ASC, \"auth_permission\".\"codename\" ASC; args=(11, 12)
DEBUG 2019-05-28 16:06:02,851 django.db.backends: (0.001) SELECT \"django_content_type\".\"id\", \"django_content_type\".\"app_label\", \"django_content_type\".\"model\" FROM \"django_content_type\" WHERE \"django_content_type\".\"app_label\" = 'guardian'; args=('guardian',)
DEBUG 2019-05-28 16:06:02,853 django.db.backends: (0.001) SELECT \"accounts_user\".\"id\", \"accounts_user\".\"password\", \"accounts_user\".\"is_superuser\", \"accounts_user\".\"email\", \"accounts_user\".\"full_name\", \"accounts_user\".\"institution\", \"accounts_user\".\"is_staff\", \"accounts_user\".\"is_active\", \"accounts_user\".\"date_joined\", \"accounts_user\".\"last_login\", \"accounts_user\".\"receive_emails\" FROM \"accounts_user\" WHERE \"accounts_user\".\"email\" = 'AnonymousUser'; args=('AnonymousUser',)
DEBUG 2019-05-28 16:06:02,855 django.db.backends: (0.001) INSERT INTO \"accounts_user\" (\"password\", \"is_superuser\", \"email\", \"full_name\", \"institution\", \"is_staff\", \"is_active\", \"date_joined\", \"last_login\", \"receive_emails\") VALUES ('!aLtoNaPtWFabXgqPfdYzzWXDLXqA2dvRojwahs89', false, 'AnonymousUser', '', '', false, true, '2019-05-28T16:06:02.853637+00:00'::timestamptz, NULL, false) RETURNING \"accounts_user\".\"id\"; args=('!aLtoNaPtWFabXgqPfdYzzWXDLXqA2dvRojwahs89', False, 'AnonymousUser', '', '', False, True, datetime.datetime(2019, 5, 28, 16, 6, 2, 853637, tzinfo=<UTC>), None, False)
DEBUG 2019-05-28 16:06:02,857 django.db.backends: (0.001) SELECT \"accounts_user\".\"email\" FROM \"accounts_user\" WHERE (\"accounts_user\".\"is_superuser\" = true AND NOT (\"accounts_user\".\"id\" = 5)); args=(True, 5)

:stderr: Traceback (most recent call last):
  File \"/opt/django/venv/lib/python3.5/site-packages/guardian/management/__init__.py\", line 33, in create_anonymous_user
    User.objects.get(**lookup)
  File \"/opt/django/venv/lib/python3.5/site-packages/django/db/models/manager.py\", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File \"/opt/django/venv/lib/python3.5/site-packages/django/db/models/query.py\", line 380, in get
    self.model._meta.object_name
accounts.models.DoesNotExist: User matching query does not exist.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File \"./manage.py\", line 23, in <module>
    execute_from_command_line(sys.argv)
  File \"/opt/django/venv/lib/python3.5/site-packages/django/core/management/__init__.py\", line 364, in execute_from_command_line
    utility.execute()
  File \"/opt/django/venv/lib/python3.5/site-packages/django/core/management/__init__.py\", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File \"/opt/django/venv/lib/python3.5/site-packages/django/core/management/base.py\", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File \"/opt/django/venv/lib/python3.5/site-packages/django/core/management/base.py\", line 330, in execute
    output = self.handle(*args, **options)
  File \"/opt/django/venv/lib/python3.5/site-packages/django/core/management/commands/migrate.py\", line 227, in handle
    self.verbosity, self.interactive, connection.alias, apps=post_migrate_apps, plan=plan,
  File \"/opt/django/venv/lib/python3.5/site-packages/django/core/management/sql.py\", line 53, in emit_post_migrate_signal
    **kwargs
  File \"/opt/django/venv/lib/python3.5/site-packages/django/dispatch/dispatcher.py\", line 193, in send
    for receiver in self._live_receivers(sender)
  File \"/opt/django/venv/lib/python3.5/site-packages/django/dispatch/dispatcher.py\", line 193, in <listcomp>
    for receiver in self._live_receivers(sender)
  File \"/opt/django/venv/lib/python3.5/site-packages/guardian/management/__init__.py\", line 38, in create_anonymous_user
    user.save()
  File \"/opt/django/venv/lib/python3.5/site-packages/django/contrib/auth/base_user.py\", line 80, in save
    super(AbstractBaseUser, self).save(*args, **kwargs)
  File \"/opt/django/venv/lib/python3.5/site-packages/django/db/models/base.py\", line 808, in save
    force_update=force_update, update_fields=update_fields)
  File \"/opt/django/venv/lib/python3.5/site-packages/django/db/models/base.py\", line 848, in save_base
    update_fields=update_fields, raw=raw, using=using,
  File \"/opt/django/venv/lib/python3.5/site-packages/django/dispatch/dispatcher.py\", line 193, in send
    for receiver in self._live_receivers(sender)
  File \"/opt/django/venv/lib/python3.5/site-packages/django/dispatch/dispatcher.py\", line 193, in <listcomp>
    for receiver in self._live_receivers(sender)
  File \"/opt/django/WebLab/weblab/accounts/signals.py\", line 6, in user_created
    send_user_creation_email(instance)
  File \"/opt/django/WebLab/weblab/accounts/emails.py\", line 27, in send_user_creation_email
    admin_emails,
  File \"/opt/django/venv/lib/python3.5/site-packages/django/core/mail/__init__.py\", line 62, in send_mail
    return mail.send()
  File \"/opt/django/venv/lib/python3.5/site-packages/django/core/mail/message.py\", line 348, in send
    return self.get_connection(fail_silently).send_messages([self])
  File \"/opt/django/venv/lib/python3.5/site-packages/django/core/mail/backends/smtp.py\", line 104, in send_messages
    new_conn_created = self.open()
  File \"/opt/django/venv/lib/python3.5/site-packages/django/core/mail/backends/smtp.py\", line 64, in open
    self.connection = self.connection_class(self.host, self.port, **connection_params)
  File \"/usr/lib/python3.5/smtplib.py\", line 251, in __init__
    (code, msg) = self.connect(host, port)
  File \"/usr/lib/python3.5/smtplib.py\", line 335, in connect
    self.sock = self._get_socket(host, port, self.timeout)
  File \"/usr/lib/python3.5/smtplib.py\", line 306, in _get_socket
    self.source_address)
  File \"/usr/lib/python3.5/socket.py\", line 711, in create_connection
    raise err
  File \"/usr/lib/python3.5/socket.py\", line 702, in create_connection
    sock.connect(sa)
OSError: [Errno 101] Network is unreachable
```

**Comment by MichaelClerx** [comment]
Guessing that's because the anonymous user didn't exist. Seems quite easy to make this fall over though

**Comment by MichaelClerx** [comment]
Trying a second time fixes that :D

**Comment by MichaelClerx** [comment]
OK. Now for an actual issue with this:

1. Create a model X, make it public
2. Create a new version of X, make it private
3. Log in as a different user, X is visible in the public view
4. Click X, it shows you the old public version
5. Click "list all versions"

```
Page not found
Request url: http://localhost:8088/entities/models/3/versions/
Raised by: entities.views.EntityVersionListView
```

**Pull request merged**